### DISCUSSION TITLE
Does minio support s3 bucket policy?

### DISCUSSION THREAD

**Discussion by adux6991** [minio/minio#17412] [Q&A] [answered]
AWS supports [bucket policy](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-policies.html), which is attached to a specific bucket and can be used to share a bucket to other users.

## Expected Behavior
Minio supports s3 bucket policy.

## Current Behavior
Bucket policy not effective in Minio.

## Possible Solution
Please clarify in docs whether or not minio supports bucket policy, since this is part of "s3-compatible".

## Steps to Reproduce (for bugs)

1. Install Minio Operator in K8s
2. Create a Minio tenant
3. Create two users: user1 and user2
4. Create a bucket: bucket1
5. Create a policy: bucket1-rw
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:DeleteObject",
                "s3:GetBucketPolicy",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:PutBucketPolicy",
                "s3:PutObject"
            ],
            "Resource": [
                "arn:aws:s3:::bucket1",
                "arn:aws:s3:::bucket1/*"
            ]
        }
    ]
}
```
6. Attach policy bucket1-rw to user1
7. user1 associate a bucket policy to bucket1 using s3cmd, which grants read/write permission of bucket1 to user2
```bash
$ cat bucket1.policy
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": ["s3:ListBucket", "s3:GetObject", "s3:PutObject", "s3:DeleteObject"],
            "Resource": [
                "arn:aws:s3:::bucket1",
                "arn:aws:s3:::bucket1/*"
            ],
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "arn:aws:iam:::user/user2"
                ]
            }
        }
    ]
}

$ s3cmd -c user1.cfg setpolicy ./bucket1.policy s3://bucket1
s3://bucket1/: Policy updated
```
8. user2 tries to read bucket1 but failed
```bash
$ s3cmd -c user2.cfg ls s3://bucket1
ERROR: Access to bucket 'bucket1' was denied
ERROR: S3 error: 403 (AccessDenied): Access Denied.
```
9. replace the `arn:aws:iam:::user/user2` part in bucket1.policy to `*` and try again, still failed
```bash
$ cat bucket1.policy
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": ["s3:ListBucket", "s3:GetObject", "s3:PutObject", "s3:DeleteObject"],
            "Resource": [
                "arn:aws:s3:::bucket1",
                "arn:aws:s3:::bucket1/*"
            ],
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            }
        }
    ]
}

$ s3cmd -c user1.cfg setpolicy ./bucket1.policy s3://bucket1
s3://bucket1/: Policy updated

$ s3cmd -c user2.cfg ls s3://bucket1
ERROR: Access to bucket 'bucket1' was denied
ERROR: S3 error: 403 (AccessDenied): Access Denied.
```

## Your Environment

* Version used (`minio --version`):
minio version RELEASE.2023-05-04T21-44-30Z (commit-id=5569acd95cbc8c7d3fac59cf2aba77baa1b9bbeb)
Runtime: go1.20.3 linux/amd64
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Copyright: 2015-2023 MinIO, Inc.
* Server setup and configuration:
Install Minio Operator in K8s and create a Minio tenant using default configuration.
* Operating System and version (`uname -a`):
Darwin MacBook-Pro.local 21.6.0 Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64 x86_64 i386 Darwin

**Comment by harshavardhana** [comment, marked as answer]
We do support bucket policies for only anonymous access. Bucket policies are not assigned to any user or you can't grant access to users via bucket policies. For that you need to use IAM policies.
